### PR TITLE
Implement script component and processor

### LINKED ISSUE #283: Implement script processor
**Feature:** [Execute script tags found in content](https://community.frontity.org/t/execute-script-tags-found-in-content/1060)
**User Story**: Script tags included in the content should be executed.

- [x] Create the script processor.
- [x] Unit tests.
- [x] Update docs.

Branch: `script-processor`
[Sprint Card](https://github.com/orgs/frontity/projects/2#card-32044424)

### LINKED ISSUE #284: Implement script component
**Feature:** [Execute script tags found in content](https://community.frontity.org/t/execute-script-tags-found-in-content/1060)
**User Story**: Script tags included in the content should be executed.

- [x] Create the script component.
- [x] e2e tests.
- [x] Update docs.

Branch: `script-processor`
[Sprint Card](https://github.com/orgs/frontity/projects/2#card-32044424)

### PR DESCRIPTION
**Pull request by iamuchejude**
#### Description of what you did:

Implement a script processor to execute script tags from embeds.

#### My PR is a:

- 🔝 Improvement

#### Is the PR ready to be merged?

- ✅ Yes!

#### Feature Request

https://community.frontity.org/t/execute-script-tags-found-in-content/1060

#### Docs Changes
https://app.gitbook.com/@frontity-tech/s/frontity/~/drafts/-LzGMLsfZI_8PDDtUgAv/api-reference-1/frontity-components/@drafts

----

Fixes #284.
Fixes #283.

### CODE REVIEW

**Review by luisherranz — CHANGES_REQUESTED** [review]
Looks great to me so far @iamuchejude :)

We use [`changesets`](https://github.com/atlassian/changesets) for versioning. You need to [add a changeset](https://github.com/atlassian/changesets/blob/master/docs/adding-a-changeset.md) to this PR. Take a look at the changeset docs and if you have any doubt let us know.

I've also left a comment about a change I think we should make.

Apart from that, maybe we can have a quick call tomorrow to see how to add e2e tests for this.

File: packages/components/script.tsx [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import React, { useRef, useEffect } from "react";
+import { connect } from "frontity";
+import { Connect, Package } from "frontity/types";
+
+type ScriptType = React.FC<Connect<Package, { src: string }>>;
+
+const generateId = () => Math.trunc(Math.random() * 9999999).toString();
+
+const Script: ScriptType = ({ state, ...props }) => {
+  const scriptId = useRef(generateId());
```
Comment (luisherranz) [inline packages/components/script.tsx]: Instead of `generateId` which will break hydration I would try to keep the reference to the `script`. Something like this:

```js
const Script: ScriptType = ({ state, ...props }) => {
  let script = null;
  const { src } = props;

  useEffect(() => {
    if (state.frontity.rendering === "csr") {
      script = window.document.createElement("script");
      script.src = src;
      script.async = true;
      script.charset = "utf-8";

      window.document.body.appendChild(script);
    }

    return () => {
      if (script) window.document.body.removeChild(script);
    };
  }, []);

  return <script async src={src} charSet="utf-8" />;
};
```

File: packages/html2react/processors/image.tsx [hunk 2]
```diff
@@ -23,6 +23,8 @@ const image: Processor = {
 export default image;
 
 const ContentImage: React.FC<any> = props => {
+  console.log(props);
```
Comment (luisherranz) [inline packages/html2react/processors/image.tsx]: Careful with the console logs.

File: packages/mars-theme/src/index.js [hunk 3]
```diff
@@ -44,7 +45,7 @@ const marsTheme = {
        * Add a processor to `html2react` so it processes the `<img>` tags
        * inside the content HTML. You can add your own processors too
        */
-      processors: [image]
+      processors: [image, script]
```
Comment (luisherranz) [inline packages/mars-theme/src/index.js]: Instead of adding the processor on each theme, I think we can add it by default in `html2react`:
https://github.com/frontity/frontity/blob/dev/packages/html2react/src/index.ts#L11

because we are "fixing something we broke". It should be fixed by default.

**Review by luisherranz — CHANGES_REQUESTED** [review]

File: packages/html2react/processors/script.tsx [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import { Processor, Element } from "../types";
+import Script from '@frontity/components/script';
+
+const script: Processor = {
+  test: node => node.type === "element" && node.component === "script",
```
Comment (luisherranz) [inline packages/html2react/processors/script.tsx]: We need to add a name (I guess `"script"` it's fine) and a priority, which I think it should be at least 20 so other people can overwrite this.

**Review by luisherranz — CHANGES_REQUESTED** [review]

**Comment by luisherranz** [comment]
Right now it looks like the best idea is to `return null` in the `<Script>` component because if not it will be executed/loaded twice in SSR. We need to figure out how to make it work with SSR when Frontity doesn't load.

File: examples/mars-theme-example/frontity.settings.js [hunk 5]
```diff
@@ -32,7 +32,7 @@ export default {
       name: "@frontity/wp-source",
       state: {
         source: {
-          api: "https://test.frontity.io/wp-json"
+          api: "http://frontity.local/wp-json"
```
Comment (luisherranz) [inline examples/mars-theme-example/frontity.settings.js]: You need to revert this change.

File: packages/components/script.tsx [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+type ScriptType = React.FC<
+  Connect<Package, { src?: string; code?: string; dataId?: string }>
+>;
+
+const Script: ScriptType = ({ state, ...props }) => {
+  let script = null;
+  const { src, code, dataId } = props;
+
+  useEffect(() => {
+    if (state.frontity.rendering === "csr") {
```
Comment (luisherranz) [inline packages/components/script.tsx]: I think there's not much sense to use `state.frontity.rendering` now that we are returning null.

File: packages/components/script.tsx [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+      script = window.document.createElement("script");
+
+      if (src) script.src = src;
+      if (code) eval(code);
```
Comment (luisherranz) [inline packages/components/script.tsx]: I think we don't need to add a script tag if what we are going to do is to use `eval` for the code.

File: packages/components/script.tsx [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+      if (dataId) script.id = dataId;
```
Comment (luisherranz) [inline packages/components/script.tsx]: Is there any reason to use `dataId` instead of `id` directly?

File: packages/components/script.tsx [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+      script.async = true;
+      script.charset = "utf-8";
```
Comment (luisherranz) [inline packages/components/script.tsx]: We can remove `charset`. It looks like it's been deprecated: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script

**Review by luisherranz — CHANGES_REQUESTED** [review]
Looks great @iamuchejude :)

Last things we can explore:
- Do e2e tests for the processor.
- Pass down other `script` props, like `type`.

**Comment by luisherranz** [comment]
Let's only apply the processor if type is missing or it's javascript (text/javascript, application/javascript...).

**Review by luisherranz — APPROVED** [review]
Great work @iamuchejude, ready for merge :)

Is the code finished?

**Comment by luisherranz** [comment]
I've merged `dev` to solve a problem with the cypress library not working in Ubuntu 18.04.

**Comment by iamuchejude** [comment]
@luisherranz The PR is ready to be merged. :)

**Comment by luisherranz** [comment]
Awesome. Thanks Jude!

If we merge this PR it will close the related issues and the "update docs" tasks are not finished yet so we need to figure out how to wait and review the documentation. I'm going to wait until I talk with @SantosGuillamot about this.

**Comment by iamuchejude** [comment]
Thanks, Luis! 🙌

I'd have a meeting with @SantosGuillamot on Monday to talk about documentation. I think that's all that is left for the issues to be resolved and closed.

**Review by luisherranz — CHANGES_REQUESTED** [review]
Great work @iamuchejude.

- I've added an e2e test to check that the script is actually being loaded.
- I've moved the react testing library dependency to `devDependencies`.
- I've added a unit test to check if the valid types are being picked up fine, but it fails. Can you check it out?

I'll review the docs next 👍

**Comment by iamuchejude** [comment]
Okay @luisherranz

**Review by luisherranz — CHANGES_REQUESTED** [review]
Docs are great, thanks. Could you please add a couple of examples of usage of the `<Script>` component? One with `src` and another with the `code` prop. Thanks! :)

**Review by luisherranz — APPROVED** [review]
Thanks @iamuchejude!

I've added TypeScript props to the Html2react processor but other than that it's fine, thanks!

Be careful with committing the lerna logs and package-lock files (we must add those to the .gitignore file).

**Comment by luisherranz** [comment]
One test is failing. I'll fix it later.

**Comment by luisherranz** [comment]
Fixed. `script` was missing because I accidentally removed it when I merged with the latest `dev` branch.
